### PR TITLE
Add matches model and related functionality

### LINKED ISSUE #9: Get current match of the week

### LINKED ISSUE #13: Create match and get all matches
This consists of creating the `Match` model, which includes a model, view, serializers, and controllers. Additionally, this includes populating the endpoint `/api/matches/`.

### LINKED ISSUE #14: Get match by ID
This includes a view, serializer, and endpoint population for individual matches. The endpoint is `/api/matches/<match_id>/`.

### PR DESCRIPTION
**Pull request by noah-solomon**
## Overview

- Added match model, view, serializer, controllers, and endpoints

### **GET `/api/matches/`**

Required Headers:
`Authorization`: `Token access_token`

**200 OK**

Response Body

```json
{
    "success": true,
    "data": [
        {
            "id": 1,
            "status": "created",
            "users": [
                {
                    "id": 1,
                    "net_id": "njs99",
                    "first_name": "Noah",
                    ...
                },
            ]
        },
    ]
}
```

### **GET `/api/matches/<id>/`**

Required Headers:
`Authorization`: `Token access_token`

**200 OK**

Response Body

```json
{
    "success": true,
    "data": {
        "id": 1,
        "status": "created",
        "users": [
            {
                "id": 1,
                "net_id": "njs99",
                "first_name": "Noah",
                    ...
            },
        ]
    }
}
```

**404 NOT FOUND**

Response Body

```json
{
    "success": false,
    "error": "Match does not exist"
}
```

### **GET `/api/matches/current/`**
Required Headers:
`Authorization`: `Token access_token`

**200 OK**

Response Body (returns the other user that you're matched with)

```json
{
    "success": true,
    "data": {
        "id": 1,
        "status": "active",
        "matched_user": {
            "id": 1,
            "net_id": "njs99",
            "first_name": "Noah",
            ...
        },
        "proposer_id": 1,
        "accepted_ids": [1, 2],
        "proposed_meeting_times": null,
        "proposed_locations": null,
        "meeting_location": 2,
        "meeting_time": "2021-04-21 15:53:00.819416"
    }
}
```
---

### **POST `/api/matches/`**

Required Headers:
`Authorization`: `Token access_token`

**200 OK**

Request Body - a match that already exists and has status: created

```json
{
    "ids": [1, 2] 
}
```

Response Body

```json
{
    "success": true
}
```

**201 CREATED**

Request Body - a match that does not yet exist, or a match that exists but without status: created (i.e. a previous match)

```json
{
    "ids": [1, 2] 
}
```

Response Body

```json
{
    "success": true
}
```

**404 NOT FOUND**

Request Body

```json
{
    "ids": [1, 4]
}
```

Response Body

```json
{
    "success": false,
    "error": "User with id 4 does not exist"
}
```

**400 BAD REQUEST**

Request Body (must include `ids`)

```json
{
    "bad key": "asdf"
}
```

Response Body

```json
{
    "success": false,
    "error": "POST body is misformatted"
}
```

### **POST `/api/matches/<id>/`**

Required Headers:
`Authorization`: `Token access_token`

**200 OK**

Request Body (can update whichever fields you choose, we will change status based on what you update)

```json
{
    "proposed_meeting_times": ["0,1","10.5,2,3","3"],
    "proposed_locations": [1],
    OR
    "meeting_location": 2,
    "meeting_time": "2021-04-21 15:53:00.819416"
}
```

Response Body

```json
{
    "success": true,
    "data": {
        "id": 1,
        "status": "created",
        "users": [
            {
                "id": 1,
                "net_id": "njs99",
                "first_name": "Noah",
                    ...
            },
        ]
    }
}
```

---

### **DELETE `/api/matches/<id>/`**

Required Headers:
`Authorization`: `Token access_token`

**200 OK**

Response Body

```json
{
    "success": true
}
```

## Changes Made

- Created matches model with all relationships: proposed locations, meeting location, and users
- Created validators for client input to create pseudo-arrays in Django
- Created serializers that either return both users in a match or only the other user that you are matched with (based on current user's authentication)
- Created controllers to create/update matches

## Test Coverage

Postman testing

## Next Steps (delete if not applicable)

Matching algorithm

## Related PRs or Issues (delete if not applicable)

- Closes #9 
- Closes #13 
- Closes #14

### CODE REVIEW

File: src/match/controllers/update_match_controller.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        if proposed_locations is not None:
+            output_messages.append(self._update_proposed_locations(proposed_locations))
+        if meeting_location is not None:
+            output_messages.append(self._update_meeting_location(meeting_location))
+        self._validate_match_status()
+        for error in output_messages:
+            if error is not None:
+                return error
+        self._match.save()
+        return success_response()
```
Comment (alanna-zhou) [inline src/match/controllers/update_match_controller.py:52]: if this is just returning a `success_response` you should make this into an asynchronous task -- you should make an issue for this and make another PR for it tho

File: src/match/match_status.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+CREATED = "created"
+PROPOSED = "proposed"
+ACTIVE = "active"
```
Comment (alanna-zhou) [inline src/match/match_status.py:3]: love to see it

File: src/match/models.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
+from django.contrib.auth.models import User
+from django.db import models
+from location.models import Location
+from match.validators import validate_int_list
+from match.validators import validate_times_list
+
+
+class Match(models.Model):
+    status = models.CharField(max_length=20)
```
Comment (alanna-zhou) [inline src/match/models.py:9]: good, always try to use CharField over TextField

File: src/match/models.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    user_1 = models.ForeignKey(
+        User, default=None, null=True, on_delete=models.CASCADE, related_name="+"
```
Comment (alanna-zhou) [inline src/match/models.py:11]: the related name should be `match` no?

File: src/match/models.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    )
+    user_2 = models.ForeignKey(
+        User, default=None, null=True, on_delete=models.CASCADE, related_name="+"
```
Comment (alanna-zhou) [inline src/match/models.py:14]: related name shouldn't just be a `+`

the related name is like if you found a `User` object and tried to access the `match` 
notice how i can access `User` through `match` with `user_2`
so then the related object must be of `match` type right

File: src/match/models.py (hunk 4 above)
Comment (alanna-zhou) [reply src/match/models.py:11]: or `proposed_match` or `accepted_match` -- it needs to make sense in your context and im not sure what that is

File: src/match/models.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        max_length=10,
+    )
+    proposed_meeting_times = models.TextField(
+        default="[]", null=True, validators=[validate_times_list]
+    )
+    proposed_locations = models.ManyToManyField(
+        Location, default=None, related_name="+"
+    )
+    meeting_location = models.ForeignKey(
+        Location, default=None, null=True, on_delete=models.SET_NULL, related_name="+"
```
Comment (alanna-zhou) [inline src/match/models.py:30]: all related names need to be updated

File: src/match/serializers.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    accepted_ids = serializers.SerializerMethodField("get_accepted_ids")
+    proposed_meeting_times = serializers.SerializerMethodField(
+        "get_proposed_meeting_times"
+    )
+    proposed_locations = serializers.SerializerMethodField("get_proposed_locations")
+    meeting_location = serializers.SerializerMethodField("get_meeting_location")
+
+    def __init__(self, *args, **kwargs):
+        self.request_user = kwargs.pop("user")
+        super().__init__(*args, **kwargs)
```
Comment (alanna-zhou) [inline src/match/serializers.py:20]: wait why do you need this im very confused

this is just a serializer

Comment (alanna-zhou) [reply src/match/serializers.py:20]: maybe im high

Comment (alanna-zhou) [reply src/match/serializers.py:20]: and i dont actually know hwat im talking abt

File: src/match/serializers.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+    def get_matched_user(self, match):
+        if self.request_user.id == match.user_1.id:
```
Comment (alanna-zhou) [inline src/match/serializers.py:23]: you can just check `if self.request_user == match.user_1`

File: src/match/validators.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
+from django.core.exceptions import ValidationError
+from django.utils.translation import gettext_lazy as _
+
+
+def validate_times_list(value):
```
Comment (alanna-zhou) [inline src/match/validators.py:5]: by times do you mean like timestamps? or like availabilities? might be helpful to write a docstring here or sth

File: src/match/validators.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+            updated_list.append(str(nested_list)[1:-1])
+        except (ValueError, TypeError):
+            raise ValidationError(
+                _(f"{value} is not a valid list of times"),
+                params={"value": value},
+            )
+    value = updated_list
+
+
+def validate_int_list(value):
```
Comment (alanna-zhou) [inline src/match/validators.py:22]: `value` is not a good name since you're assuming it's a list

why not `items`

Comment (alanna-zhou) [inline src/match/validators.py:22]: wait are you even validating or are you converting the list to an int list 

im not sure why this is needed, this is better: https://stackoverflow.com/questions/6009589/how-to-test-if-every-item-in-a-list-of-type-int

File: src/match/views.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+
+class CurrentMatchView(generics.GenericAPIView):
+    serializer_class = MatchSerializer
+    permission_classes = api_settings.CONSUMER_PERMISSIONS
+
+    def get(self, request):
+        """Get current match by user."""
+        match = Match.objects.filter(
+            Q(user_1=request.user.id) | Q(user_2=request.user.id)
```
Comment (alanna-zhou) [inline src/match/views.py:69]: why are you checking user id? shouldn't you be checking `user_1=request.user` ? how does this not yell at you lol

File: src/person/models.py [hunk 12]
```diff
@@ -13,3 +17,11 @@ class Person(models.Model):
...
     graduation_year = models.CharField(max_length=4, default=None, null=True)
     pronouns = models.CharField(max_length=20, default=None, null=True)
+    goals = models.TextField(default=None, null=True)
+    talking_points = models.TextField(default=None, null=True)
+    matches = models.ManyToManyField(Match, default=None)
+    availability = models.TextField(default=None, null=True)
+    locations = models.ManyToManyField(Location, default=None)
+    interests = models.ManyToManyField(Interest, default=None)
+    groups = models.ManyToManyField(Group, default=None)
+    onboarded = models.BooleanField(default=None, null=True)
```
Comment (alanna-zhou) [inline src/person/models.py:27]: since it's a boolean, probably `has_onboarded` is better

**Review by alanna-zhou — CHANGES_REQUESTED** [review]
nice job writing all of this!!! you're so close!! just some comments to address before i approve

File: src/match/serializers.py (hunk 7 above)
Comment (noah-solomon) [reply src/match/serializers.py:20]: @alanna-zhou i added it to pass in the current user into the serializer so that the serializer returns the correct `matched_user` since frontend only wants us to return that

File: src/match/serializers.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+        proposed_meeting_times = json.loads(
+            match.proposed_meeting_times.replace("'", '"')
+        )
+        if proposed_meeting_times == []:
+            return None
+        return proposed_meeting_times
+
+    def get_proposed_locations(self, match):
+        proposed_locations = []
+        for location in match.proposed_locations.all():
```
Comment (chalo2000) [inline src/match/serializers.py:48]: this .all() thing again, am i crazy and this works?

File: src/match/serializers.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+            "accepted_ids",
+            "proposed_meeting_times",
+            "proposed_locations",
+            "meeting_location",
+            "meeting_time",
+        )
+        read_only_fields = fields
+
+
+class BothUsersMatchSerializer(serializers.ModelSerializer):
```
Comment (chalo2000) [inline src/match/serializers.py:76]: why do we need two serializers?

File: src/match/validators.py [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+            for index in range(len(nested_list)):
+                if nested_list[index] != "":
+                    nested_list[index] = float(nested_list[index])
+            updated_list.append(str(nested_list)[1:-1])
+        except (ValueError, TypeError):
+            raise ValidationError(
+                _(f"{value} is not a valid list of times"),
+                params={"value": value},
+            )
+    value = updated_list
```
Comment (chalo2000) [inline src/match/validators.py:19]: does it matter to even set value equal to the updated_list? if you needed to set the original value to updated_list you should've returned here because it looks like value is a string and this equals doesn't update the original value or help with validation at all (you're done when you get here)

File: src/match/validators.py (hunk 10 above)
Comment (chalo2000) [reply src/match/validators.py:22]: ^ rekt, you can do it in a single line

also @alanna-zhou i think this is him trying to create an int list and if it fails for some reason then that means there is something that is not an int in the list, meaning the list is bad... but i'm not sure why we even need to validate this because this is being referenced in the `UpdateMatchController` which should only be called internally to obscure away the details of the implementation from the frontend so they only need to call a route and we take care of the rest

File: src/match/views.py [hunk 16]
```diff
@@ -0,0 +1,81 @@
+import json
+
+from api import settings as api_settings
+from api.utils import failure_response
+from api.utils import success_response
+from django.db.models import Q
```
Comment (chalo2000) [inline src/match/views.py:6]: what's this for?

File: src/match/views.py [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+
+    def get(self, request):
+        """Get all matches."""
+        matches = Match.objects.all()
+        return success_response(self.serializer_class(matches, many=True).data)
+
+    def post(self, request):
+        """Create a match."""
+        try:
+            data = json.loads(request.body)
```
Comment (chalo2000) [inline src/match/views.py:28]: this is what I was referring to about the json loading for the request being redundant in `CreateMatchController` meaning that it's pointless to pass `request` into the controller

**Review by chalo2000 — CHANGES_REQUESTED** [review]
wow I wasn't even aware you were working on my issues, we have a lot to talk about to make sure this works as intended but this is a good foundation so far :)

File: src/match/validators.py (hunk 10 above)
Comment (alanna-zhou) [reply src/match/validators.py:22]: yeah if this is just to validate frontend's request i'd just say forget about it. it wastes precious time

Comment (noah-solomon) [reply src/match/validators.py:22]: the reason I have this is so that frontend sends us the correctly-formatted lists when sending `proposed_locations` and `proposed_meeting_times` but since you guys think we don't need to validate those I'll just take this out

File: src/match/controllers/update_match_controller.py [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+
+    def _modify_attribute(self, attr_name, attr_value):
+        """Modify an attribute if it isn't None and has been changed."""
+        if attr_value is not None and attr_value != getattr(self._match, attr_name):
+            setattr(self._match, attr_name, attr_value)
+
+    def _validate_acceptors_and_times(self, body_accepted_ids, body_proposed_times):
```
[the remark this answers is not in the document]
Comment (noah-solomon) [reply src/match/controllers/update_match_controller.py:59]: yea i'm just gonna scrap this lol

File: src/match/controllers/update_match_controller.py [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+                return failure_response(
+                    f"Location {new_location_id} does not exist",
+                    status.HTTP_400_BAD_REQUEST,
+                )
+            new_location = Location.objects.filter(id=new_location_id)
+            if not new_location:
+                return failure_response(
+                    f"Location with id {new_location_id} does not exist."
+                )
+            if new_location[0] not in self._match.proposed_locations.all():
```
[the remark this answers is not in the document]
Comment (noah-solomon) [reply src/match/controllers/update_match_controller.py:107]: already talked to gonzalo but for anyone looking at this: `all()` is a django function for queryset objects

File: src/match/controllers/update_match_controller.py [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+        if not self._match:
+            return failure_response("Match does not exist.")
+        self._match = self._match[0]
+        proposed_meeting_times = self._data.get("proposed_meeting_times")
+        proposed_locations = self._data.get("proposed_locations")
+        meeting_location = self._data.get("meeting_location")
+        meeting_time = self._data.get("meeting_time")
+
+        # Next, modify the attributes that may have changed
+        error_responses = []
```
Comment (alanna-zhou) [inline src/match/controllers/update_match_controller.py:31]: wait why are you doing this to error responses

you should just return a failure response at the very first time you encounter one

File: src/match/controllers/update_match_controller.py [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+        Example: ["", "", "9.5", "", "", "", ""] turns into
+        '2021-05-11 09:30:00' i.e. the Tuesday after now() at 9:30 am
+        """
+        # simple_meeting is [weekday, time] from times param
+        simple_meeting = [0, 0]
+        for index in range(len(times)):
+            if times[index] != "":
+                simple_meeting = index, float(times[index])
+        now = datetime.datetime.now()
+        hour, minute = divmod(simple_meeting[1], 1)
```
Comment (alanna-zhou) [inline src/match/controllers/update_match_controller.py:97]: what does divmod do?

File: src/match/controllers/update_match_controller.py [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+                    "Match has already been accepted", status.HTTP_400_BAD_REQUEST
+                )
+            self._match.status = match_status.ACTIVE
+            # use json.loads to convert a string of a list to an actual list
+            accepted_ids = json.loads(self._match.accepted_ids)
+            if self._user.id not in accepted_ids:
+                accepted_ids.append(self._user.id)
+                self._match.accepted_ids = accepted_ids
+            self._match.proposed_meeting_times = None
+            self._match.proposed_locations.set([])
```
Comment (alanna-zhou) [inline src/match/controllers/update_match_controller.py:131]: why do you need to do this? can't you just leave `self._match.proposed_locations` alone?

File: src/match/models.py [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+        default=None,
+        null=True,
+        on_delete=models.SET_NULL,
+        related_name="matches_2",
+    )
+    proposer_id = models.IntegerField(default=None, null=True)
+    accepted_ids = models.CharField(
+        default=None,
+        null=True,
+        validators=[validate_int_list],
```
Comment (alanna-zhou) [inline src/match/models.py:28]: i thought you got rid of `validate_int_list`

File: src/match/serializers.py [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+    meeting_location = serializers.SerializerMethodField("get_meeting_location")
+
+    def __init__(self, *args, **kwargs):
+        self.request_user = kwargs.pop("user")
+        super().__init__(*args, **kwargs)
+
+    def get_matched_user(self, match):
+        if self.request_user == match.user_1:
+            return SimpleUserSerializer(match.user_2).data
+        elif self.request_user.id == match.user_2.id:
```
Comment (alanna-zhou) [inline src/match/serializers.py:25]: you should do `self.request_user == ...`

File: src/match/serializers.py [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+            return SimpleUserSerializer(match.user_1).data
+
+    def get_accepted_ids(self, match):
+        if match.accepted_ids is None:
+            return None
+        accepted_ids = json.loads(match.accepted_ids)
+        if accepted_ids == []:
+            return None
+        return accepted_ids
```
Comment (alanna-zhou) [inline src/match/serializers.py:34]: can't you do `return accepted_ids or None`? you should test this out in the python interpreter first, or you can enter the shell with `python manage.py shell`

File: src/match/serializers.py [hunk 26]
```diff
@@ -0,0 +1,137 @@
...
+
+    def get_proposed_meeting_times(self, match):
+        if match.proposed_meeting_times is None:
+            return None
+        # We have to replace all single quotes with double quotes for JSON
```
Comment (alanna-zhou) [inline src/match/serializers.py:39]: wait whatttt i have never seen this before haha

File: src/match/serializers.py [hunk 27]
```diff
@@ -0,0 +1,137 @@
...
+        "get_proposed_meeting_times"
+    )
+    proposed_locations = serializers.SerializerMethodField("get_proposed_locations")
+    meeting_location = serializers.SerializerMethodField("get_meeting_location")
+
+    def get_users(self, match):
+        return (
+            SimpleUserSerializer(match.user_1).data,
+            SimpleUserSerializer(match.user_2).data,
+        )
```
Comment (alanna-zhou) [inline src/match/serializers.py:90]: wait can't you just do `Serializer(..., many=True).data`

File: src/match/serializers.py [hunk 28]
```diff
@@ -0,0 +1,137 @@
...
+
+    def get_accepted_ids(self, match):
+        if match.accepted_ids is None:
+            return None
+        accepted_ids = json.loads(match.accepted_ids)
+        if accepted_ids == []:
+            return None
+        return accepted_ids
```
Comment (alanna-zhou) [inline src/match/serializers.py:98]: `return accepted_ids or None` ?

File: src/match/serializers.py [hunk 29]
```diff
@@ -0,0 +1,137 @@
...
+            return None
+        return proposed_meeting_times
+
+    def get_proposed_locations(self, match):
+        proposed_locations = []
+        for location in match.proposed_locations.all():
+            proposed_locations.append(LocationSerializer(location).data)
+        if proposed_locations == []:
+            return None
+        return proposed_locations
```
Comment (alanna-zhou) [inline src/match/serializers.py:116]: just do `LocationSerializer(match.proposed_locations.all(), many=True).data`

File: src/match/serializers.py [hunk 30]
```diff
@@ -0,0 +1,137 @@
...
+
+    def get_meeting_location(self, match):
+        meeting_location = LocationSerializer(match.meeting_location).data
+        if meeting_location == {}:
```
Comment (alanna-zhou) [inline src/match/serializers.py:120]: wait why does this matter

File: src/match/validators.py [hunk 31]
```diff
@@ -0,0 +1,25 @@
...
+            )
+
+
+def validate_int_list(items):
+    """Precondition: items is a list"""
+    if not all(isinstance(element, int) for element in items):
+        raise ValidationError(
+            _(f"{items} is not an int list"),
+            params={"items": items},
+        )
```
Comment (alanna-zhou) [inline src/match/validators.py:25]: i thought you intended to delete this

File: src/match/views.py [hunk 32]
```diff
@@ -0,0 +1,122 @@
...
+class MatchView(generics.GenericAPIView):
+    serializer_class = BothUsersMatchSerializer
+    permission_classes = api_settings.CONSUMER_PERMISSIONS
+
+    def get(self, request, id):
+        """Get match by id."""
+        match = Match.objects.filter(id=id)
+        if match:
+            return success_response(self.serializer_class(match[0]).data)
+        return failure_response("Match does not exist")
```
Comment (alanna-zhou) [inline src/match/views.py:44]: bruh just return the match lol

Comment (alanna-zhou) [reply src/match/views.py:44]: also just do `Match.objects.get(id=id)` 
if it fails, it fails

File: src/person/serializers.py [hunk 33]
```diff
@@ -20,13 +23,22 @@ def get_access_token(self, instance):
...
     instagram_username = serializers.CharField(source="person.instagram_username")
     graduation_year = serializers.CharField(source="person.graduation_year")
     pronouns = serializers.CharField(source="person.pronouns")
+    matches = serializers.SerializerMethodField("get_match_history")
+
+    def get_match_history(self, user):
+        matches = Match.objects.filter(Q(user_1=user) | Q(user_2=user)).order_by(
+            "-created_date"
+        )
+        return MatchSerializer(matches, user=user, many=True).data
```
Comment (alanna-zhou) [inline src/person/serializers.py:41]: nice!

**Review by alanna-zhou — CHANGES_REQUESTED** [review]
sorry i am requesting again 😂  

you're doing great though, a thicc PR once again!!

File: src/match/controllers/update_match_controller.py (hunk 20 above)
Comment (noah-solomon) [reply src/match/controllers/update_match_controller.py:31]: lmk if you would rather this all be in one method or if u think there's a better way

File: src/match/controllers/update_match_controller.py (hunk 22 above)
Comment (noah-solomon) [reply src/match/controllers/update_match_controller.py:131]: in our original call with frontend you mentioned that when state changes we should update non-necessary fields to None or empty set, lmk if I don't need to do that tho

File: src/match/models.py (hunk 23 above)
Comment (noah-solomon) [reply src/match/models.py:28]: oops yes i'll do that

File: src/match/serializers.py (hunk 30 above)
Comment (noah-solomon) [reply src/match/serializers.py:120]: cuz if meeting location is none in django it's an empty python set but i still wanna just return `None`, but I'm gonna do ur cool `return meeting_location or None`

File: src/match/views.py (hunk 32 above)
Comment (noah-solomon) [reply src/match/views.py:44]: wouldn't we wanna return a 404 if client is trying to get a match that doesn't exist? that's what we do for all our other models

File: src/match/controllers/create_match_controller.py [hunk 34]
```diff
@@ -0,0 +1,46 @@
...
+from rest_framework import status
+
+
+class CreateMatchController:
+    def __init__(self, data, serializer):
+        self._data = data
+        self._serializer = serializer
+
+    def process(self):
+        """Process a request to create a match."""
```
Comment (chalo2000) [inline src/match/controllers/create_match_controller.py:15]: This is obvious from the name of the controller, just remove the comment.

File: src/match/controllers/create_match_controller.py [hunk 35]
```diff
@@ -0,0 +1,46 @@
...
+        match_ids = self._data.get("ids")
+        if match_ids is None:
+            return failure_response(
+                "POST body is misformatted", status.HTTP_400_BAD_REQUEST
+            )
+        user_1 = User.objects.filter(id=match_ids[0])
+        user_2 = User.objects.filter(id=match_ids[1])
+        if not user_1:
```
Comment (chalo2000) [inline src/match/controllers/create_match_controller.py:23]: Check if not `user_1` before filtering for `user_2` so that we immediately send a failure response if `user_1` is None and don't waste computational power on the second unnecessary filter.

File: src/match/controllers/create_match_controller.py [hunk 36]
```diff
@@ -0,0 +1,46 @@
...
+            return failure_response(f"User with id {match_ids[1]} does not exist")
+        user_1 = user_1[0]
+        user_2 = user_2[0]
+        possible_match_1 = Match.objects.filter(
+            user_1=user_1, user_2=user_2, status=match_status.CREATED
+        )
+        possible_match_2 = Match.objects.filter(
+            user_1=user_2, user_2=user_1, status=match_status.CREATED
+        )
+        if possible_match_1 or possible_match_2:
```
Comment (chalo2000) [inline src/match/controllers/create_match_controller.py:35]: The top comment ESPECIALLY applies here because unlike users, we will be getting a linear increase of matches per week so if `possible_match_1` exists we can just return success immediately. 

A nice precondition for a future optimization PR that we can have is if we were to guarantee that `user_1` always has a lower `id` than `user_2` so we only need to filter through the matches a single time before creating a new one.

Also good job filtering by the status too because we want to ensure that even if someone on Pear were to theoretically meet everyone on the app that they can possibly match again with a good Pear in the future after some time period.

File: src/match/controllers/update_match_controller.py (hunk 20 above)
Comment (alanna-zhou) [reply src/match/controllers/update_match_controller.py:31]: oh hm i see... i feel like there's a better way but i won't block you. you can always refactor after you merge 👍

File: src/match/controllers/update_match_controller.py (hunk 21 above)
Comment (alanna-zhou) [reply src/match/controllers/update_match_controller.py:97]: wow that's actually very cool

File: src/match/serializers.py (hunk 30 above)
Comment (alanna-zhou) [reply src/match/serializers.py:120]: haha but don't take my word for it, i might be wrong :p 

yah im just being picky

File: src/match/views.py (hunk 32 above)
Comment (alanna-zhou) [reply src/match/views.py:44]: hmmmmmmmmm i guess 

lol you can keep this then 
i just being a picky boi

File: src/match/controllers/update_match_controller.py (hunk 20 above)
Comment (chalo2000) [reply src/match/controllers/update_match_controller.py:31]: i'd review your logic with this

if frontend gets a failure they're expecting uh oh I have an error so I need to fix it before the match is updated properly but you still do updates before sending the failure response so we can easily get into an inconsistent state (i.e. regardless of it being possible, consider the case where everything is updated except my match status because I got an error and it was sent to frontend. now my proposed locations and meeting locations will be inconsistent with the current match status which was not previously updated and more errors will occur from there, making it hard to debug)

instead of checking for error responses when updating, can you just check that the data passed in by frontend is 100% correct in a helper before proceeding with updates?

File: src/match/views.py (hunk 32 above)
Comment (chalo2000) [reply src/match/views.py:44]: little nitpick, actually put the 404 error here even though it's a default argument to be explicit like in my refactor (same for success)

**Review by chalo2000 — CHANGES_REQUESTED** [review]
getting closer to the finish line 😎  requesting changes primarily about the `error_responses` and the possible headaches that will give us when frontend is integrating matches if not fixed

File: src/match/controllers/update_match_controller.py (hunk 20 above)
Comment (noah-solomon) [reply src/match/controllers/update_match_controller.py:31]: good points, i made sure to return the failure before I ever called `match.save()` so the match won't be updated but you're right that I should catch the errors earlier

File: src/match/controllers/create_match_controller.py (hunk 36 above)
Comment (noah-solomon) [reply src/match/controllers/create_match_controller.py:35]: smart boy

File: src/match/controllers/create_match_controller.py [hunk 37]
```diff
@@ -0,0 +1,49 @@
...
+        if not user_2:
+            return failure_response(f"User with id {max(match_ids)} does not exist")
+        user_1 = user_1[0]
+        user_2 = user_2[0]
+        possible_match_1 = Match.objects.filter(
+            user_1=user_1, user_2=user_2, status=match_status.CREATED
+        )
+        if possible_match_1:
+            return success_response(None, status.HTTP_200_OK)
+        possible_match_2 = Match.objects.filter(
```
Comment (chalo2000) [inline src/match/controllers/create_match_controller.py:35]: now that you set user_1 to the user with the lowest ID and user_2 to the user with the highest ID, there is no need to check for possible_match_2 since that case is impossible

File: src/match/controllers/update_match_controller.py [hunk 38]
```diff
@@ -0,0 +1,121 @@
...
+        meeting_time = self._data.get("meeting_time")
+
+        # Next, modify the attributes that may have changed:
+        # Proposed Meeting Times
+        self._modify_attribute("proposed_meeting_times", proposed_meeting_times)
+        # Meeting Time
+        if meeting_time is not None:
+            self._update_meeting_time(meeting_time)
+        # Proposed Locations
+        if proposed_locations is not None:
```
Comment (chalo2000) [inline src/match/controllers/update_match_controller.py:37]: any reason why you go through the effort to check if a Location is already in the previous match object before adding and removing any locations whose ids were not in the proposed_location instead of just building up a new query set to assign `_match.proposed_locations` to?

File: src/match/controllers/update_match_controller.py [hunk 39]
```diff
@@ -0,0 +1,121 @@
...
+        if meeting_location_id is not None:
+            new_meeting_location = Location.objects.filter(id=meeting_location_id)
+            if not new_meeting_location:
+                return failure_response(
+                    f"Location with id {meeting_location_id} does not exist."
+                )
+            self._modify_attribute("meeting_location", new_meeting_location[0])
+
+        # Based on what changed, check for status conflicts
+        if proposed_meeting_times is not None and proposed_locations is not None:
```
Comment (chalo2000) [inline src/match/controllers/update_match_controller.py:57]: I would check these BEFORE modifying the match because then you'll have the issue of "uh oh you can't do this" but you already did it and now your match's proposed meeting times or locations are in a messed up

File: src/match/serializers.py [hunk 40]
```diff
@@ -0,0 +1,128 @@
...
+    def get_accepted_ids(self, match):
+        if match.accepted_ids is None:
+            return None
+        accepted_ids = json.loads(match.accepted_ids)
+        return accepted_ids or None
+
+    def get_proposed_meeting_times(self, match):
+        if match.proposed_meeting_times is None:
+            return None
+        proposed_meeting_times = json.loads(
```
Comment (chalo2000) [inline src/match/serializers.py:99]: this has been bothering me a lot lol can you try `json.dumps()`

https://www.educative.io/edpresso/what-is-the-difference-between-jsonloads-and-jsondumps

Screen Shot 2021-05-15 at 4 52 59 PM

**Review by chalo2000 — APPROVED** [review]
much better - approving to get this in ASAP but there is a comment that **must** be addressed before merging

**Review by alanna-zhou — APPROVED** [review]
im getting tired of looking at this so if there is any mistake here it will be on you to fix hahahahahaahahahahahaha
